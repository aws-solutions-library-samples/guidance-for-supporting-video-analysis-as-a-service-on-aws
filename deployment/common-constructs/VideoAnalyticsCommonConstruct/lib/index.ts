export { VideoAnalyticsAsyncWorkflowResource, Workflow, WorkflowProps, GsiProps, WorkflowLambdaProps } from './video_analytics_async_workflow';
export { createLambdaRole, createBasicLambdaRole } from './serviceConstructs/iamUtils';
export { createTable, createGSI } from './serviceConstructs/ddbUtils';
