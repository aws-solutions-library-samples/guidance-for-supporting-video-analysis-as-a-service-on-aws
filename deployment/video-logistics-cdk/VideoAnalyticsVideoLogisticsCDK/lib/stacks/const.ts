export const FORWARDING_RULES_BUCKET_NAME = "videoanalytics-fwd-rules";
export const FORWARDING_RULES_ACCESS_LOGS_BUCKET_NAME =
  "videoanalytics-fwd-rules-access-logs";
export const FORWARDING_RULES_SQS_QUEUE_NAME =
  "aws-ip-videoanalytics-VLControlPlane-forwardingRulesQueue";
export const FORWARDING_RULES_TABLE_NAME = "ForwardingRulesTable";
export const FORWARDING_RULES_PK_NAME = "CustomerAccountId";
export const FORWARDING_RULES_SK_NAME = "ForwardingRulesName";
export const VIDEO_EXPORT_JOB_TABLE_NAME = "VideoExportJobTable";
export const VIDEO_EXPORT_JOB_RESULT_TABLE_NAME = "VideoExportResultTable";
export const VIDEO_EXPORT_SFN_STATE_MACHINE_NAME = "VideoExportStateMachine";
export const VIDEO_EXPORT_JOB_GSI_NAME = "DeviceJobStatus-Index";
export const VIDEO_EXPORT_JOB_GSI_PK = "CustomerAccountId_DeviceId";
export const VIDEO_EXPORT_JOB_GSI_SK = "JobStatus";
export const JOB_ID = "JobId";
export const EXPORT_FILE_START_TIME = "ExportFileStartTime";
export const VIDEO_EXPORT_JOB_PARTITION_KEY = "CustomerAccountId_DeviceId";
export const JOB_STATUS = "JobStatus";
export const VIDEO_EXPORT_RESULTS_GSI_1ST_NAME = "JobIDStatus-Index";
export const VIDEO_EXPORT_RESULTS_GSI_1ST_SK = "SegmentJobStatus";
export const VIDEO_EXPORT_RESULTS_GSI_2ND_NAME = "JobIDEndTime-Index";
export const VIDEO_EXPORT_RESULTS_GSI_2ND_SK = "ExportFileEndTime";
export const VIDEO_EXPORT_ACCOUNT_JOB_ID_GSI = "AccountIDJobID-Index";
export const VIDEO_EXPORT_ACCOUNT_JOB_ID_GSI_PK = "CustomerAccountId";
// Step function state parameters
export const EXECUTION_COUNTER = "$.Counter";
export const EXPORT_RESULT_PATH = "$.exportResult";
export const EXECUTION_RESULT_IN_STATUS = "$.executionResult";
export const UPDATED_TIMESTAMP =
  "$.exportResult.Payload.newLastProcessedTimestamp";
export const IS_COMPLETE_CONDITION = "$.exportResult.Payload.isJobComplete";
export const LAST_PROCESSED_TIMESTAMP = "$.lastProcessedTimestamp";
export const WAIT_TIME_BETWEEN_LOOP = "$.exportResult.Payload.waitTime";
export const FAILED_DYNAMIC_CAUSE = "$.Cause";
// used for when execution Lambda failed cause forwarding after statusUpdate
export const FAILED_FORWARD_CAUSE = "$.Payload.Cause";
export const SHOULD_STREAM_FROM_DEVICE_PATH = "$.shouldStreamFromDevice";
export const PARTITION_KEY_PATH = "$.partitionKey";
export const SORT_KEY_PATH = "$.sortKey";
export const RESULT_PATH_ERROR = "$.error";
export const ERROR_MESSAGE_PATH = "$.error.Cause";
export const RESULT_PATH = "$.output";
export const enum PLAYBACK_SESSION_STATUS {
  FAILED = "FAILED",
  COMPLETED = "COMPLETED",
}
export const RAW_VIDEO_TIMELINE_TABLE_NAME = "RawVideoTimelineTable";
export const VIDEO_TIMELINE_TABLE_NAME = "VideoTimelineTable";
export const RAW_VIDEO_TIMELINE_PK_NAME = "CustomerId_DeviceId";
export const VIDEO_TIMELINE_PK_NAME = "CustomerId_DeviceId_TimeUnit";
export const VIDEO_TIMELINE_SK_NAME = "UnitTimestamp";
export const VIDEO_TIMELINE_TTL_ATTRIBUTE_NAME = "ExpirationTimestamp";
export const RAW_VIDEO_TIMELINE_SORT_KEY_NAME = "Timestamp";
export const TIMELINE_BUCKET_NAME = "videoanalytics-timeline-bucket";
// TODO: Update handler to match new lambda handler path
export const DENSITY_UPDATE_LAMBDA_HANDLER_PATH =
  "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.VideoDensityUpdateLambda::handleRequest";
export const EXPORT_LAMBDA_HANDLER_PATH =
  "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.VideoTimelineS3ExportLambda::handleRequest";
export const LAMBDA_SERVICE_PRINCIPAL = 'lambda.amazonaws.com';
export const LAMBDA_MANAGED_POLICY_NAME = 'service-role/AWSLambdaBasicExecutionRole';
// TODO: Update handler to match new lambda handler path
export const TIMELINE_FORWARDER_HANDLER_PATH =
  'com.amazon.awsvideoanalyticsvlcontrolplane.timeline.TimelineForwarderLambda::handleRequest';
export const LAMBDA_PACKAGE_NAME = 'AWSVideoAnalyticsVLControlPlane-1.0';
  "com.amazon.awsvideoanalyticsvlcontrolplane.timeline.TimelineForwarderLambda::handleRequest";

export const VL_ACTIVITY_JAVA_PATH_PREFIX =
  "com.amazonaws.videoanalytics.videologistics.activity";
export const VL_WORKFLOW_JAVA_PATH_PREFIX =
  "com.amazonaws.videoanalytics.videologistics.workflow";
export const LAMBDA_ASSET_PATH =
  "../../../assets/lambda-built/video-logistics-assets/VideoAnalyticsVideoLogisticsControlPlane-1.0-beta.jar";
export const OPEN_API_SPEC_PATH = "../../../assets/model/video-logistics/openapi-conversion/openapi/VideoAnalytic.openapi.json";