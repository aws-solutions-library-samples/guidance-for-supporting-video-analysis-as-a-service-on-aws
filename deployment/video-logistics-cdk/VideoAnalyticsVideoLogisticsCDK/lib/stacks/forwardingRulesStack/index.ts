// export * from './forwardingRulesStack';
export * from './opensearchStack';