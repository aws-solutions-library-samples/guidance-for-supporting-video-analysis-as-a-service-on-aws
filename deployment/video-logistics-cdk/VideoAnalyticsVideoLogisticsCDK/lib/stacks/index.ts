export * from './forwardingRulesStack';
export * from './schedulerStack';
export * from './timelineStack';
export * from './videoExportStack';
export * from './workflowStack';
export * from './serviceStack';
