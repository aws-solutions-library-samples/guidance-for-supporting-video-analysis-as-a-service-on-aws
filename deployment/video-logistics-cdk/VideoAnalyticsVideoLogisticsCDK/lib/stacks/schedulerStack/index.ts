export * from './openSearchPitCreation';
export * from './schedulerStack';