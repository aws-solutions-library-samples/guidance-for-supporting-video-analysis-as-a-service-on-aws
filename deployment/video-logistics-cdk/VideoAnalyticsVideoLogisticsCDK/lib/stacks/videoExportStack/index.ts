export * from './videoExportWorkflowResources';
export * from './videpExportStack';