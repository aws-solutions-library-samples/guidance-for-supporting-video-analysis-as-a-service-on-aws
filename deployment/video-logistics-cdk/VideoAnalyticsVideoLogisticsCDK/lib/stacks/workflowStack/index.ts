export * from './workflowResources';
export * from './bulkInferenceStack';
export * from './workflowStack';