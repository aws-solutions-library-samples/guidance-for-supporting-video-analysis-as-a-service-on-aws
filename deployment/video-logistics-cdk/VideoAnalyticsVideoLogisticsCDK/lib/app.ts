import { App } from "aws-cdk-lib";
import { 
    // ForwardingRulesStack,
    ServiceStack,
    TimelineStack,
    VideoExportStack,
    WorkflowStack
 } from "./stacks";
import { AWSRegion } from "video_analytics_common_construct";
import { VideoLogisticsBootstrapStack } from "./stacks/bootstrapStack/videoLogisticsBootstrapStack";
import { BulkInferenceStack } from "./stacks/workflowStack/bulkInferenceStack";
import { OpenSearchStack } from "./stacks/forwardingRulesStack";

const app = new App();

// Get the account and region from environment variables or use defaults
const account =
  process.env.CDK_DEPLOY_ACCOUNT ||
  process.env.CDK_DEFAULT_ACCOUNT ||
  "YOUR_DEFAULT_ACCOUNT";
const region = (process.env.CDK_DEPLOY_REGION ||
  process.env.CDK_DEFAULT_REGION ||
  "us-west-2") as AWSRegion;

// Common environment config
const envConfig = {
    env: {
        account: account,
        region: region
    },
    region: region,
    account: account
};

// Create all stacks
// new ForwardingRulesStack(app, 'VideoLogisticsForwardingRulesStack', envConfig);
// new SchedulerStack(app, 'VideoLogisticsSchedulerStack', envConfig);
// new TimelineStack(app, 'VideoLogisticsTimelineStack', envConfig);
// new VideoExportStack(app, 'VideoLogisticsVideoExportStack', envConfig);
new VideoLogisticsBootstrapStack(app, 'VideoLogisticsBootstrapStack', envConfig);
let opensearchStack = new OpenSearchStack(app, 'VideoLogisticsOpensearchStack', envConfig);
new BulkInferenceStack(app, 'VideoLogisticsBulkInferenceStack', {...envConfig, opensearchEndpoint: opensearchStack.opensearchEndpoint});
new WorkflowStack(app, 'VideoLogisticsWorkflowStack', envConfig);
new ServiceStack(app, 'VideoLogisticsServiceStack', envConfig);

app.synth();
