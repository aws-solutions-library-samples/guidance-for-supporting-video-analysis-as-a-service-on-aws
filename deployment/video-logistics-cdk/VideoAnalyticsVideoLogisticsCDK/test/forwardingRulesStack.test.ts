import { createStackSnapshotTest } from "./snapshotTestUtil";
import { ForwardingRulesStack } from "../lib/stacks";

createStackSnapshotTest("ForwardingRulesStack", ForwardingRulesStack);
