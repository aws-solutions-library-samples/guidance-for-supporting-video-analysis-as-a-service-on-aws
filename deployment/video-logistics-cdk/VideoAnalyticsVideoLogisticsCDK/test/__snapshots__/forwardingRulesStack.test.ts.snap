// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForwardingRulesStack Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DataForwarderLambda658B5D08": {
      "DependsOn": [
        "DataForwarderLambdaRoleDefaultPolicy58485E58",
        "DataForwarderLambdaRole6192FE95",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "f93e4547cbac7f4ce906b37e330f99bfb1afa04f8cffa2fd0c755adbf502c5d4.jar",
        },
        "Description": "Lambda responsible for forwarding data in accordance with forwarding rules",
        "Environment": {
          "Variables": {
            "AWSRegion": "us-east-1",
          },
        },
        "Handler": "com.amazon.awsvideoanalyticsvlcontrolplane.forwardingrules.DataForwarderLambda::handleRequest",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "DataForwarderLambdaLogGroupD801CBCE",
          },
        },
        "MemorySize": 2048,
        "Role": {
          "Fn::GetAtt": [
            "DataForwarderLambdaRole6192FE95",
            "Arn",
          ],
        },
        "Runtime": "java17",
        "Timeout": 360,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DataForwarderLambdaLogGroupD801CBCE": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "DataForwarderLambdaLogGroup",
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "DataForwarderLambdaRole6192FE95": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Allows lambda to read and delete messages from the SQS queue",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaSQSQueueExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "DataForwarderLambdaRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "DataForwarderLambdaRoleDefaultPolicy58485E58": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:List*",
                "s3:Get*",
                "s3:DeleteObject",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:aws:s3:::videoanalytics-fwd-rules-us-east-1*",
                "arn:aws:s3:::videoanalytics-fwd-rules-us-east-1*/*",
              ],
            },
            {
              "Action": "kinesis:PutRecord",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:GetRecords",
                "dynamodb:GetItem",
                "dynamodb:Query",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": [
                "*",
                {
                  "Fn::GetAtt": [
                    "ForwardingRulesTableE57C3B49",
                    "Arn",
                  ],
                },
              ],
            },
            {
              "Action": [
                "kms:Encrypt",
                "kms:Decrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DataForwarderLambdaRoleDefaultPolicy58485E58",
        "Roles": [
          {
            "Ref": "DataForwarderLambdaRole6192FE95",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventSourceMapping3449BD33": {
      "Properties": {
        "BatchSize": 10,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "ForwardingRulesQueueCCFF9CDD",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "DataForwarderLambda658B5D08",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "ForwardingRulesQueueCCFF9CDD": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "awsipvideoanalyticsVLControlPlaneforwardingRulesQueuekmsKeyA3AA2F57",
            "Arn",
          ],
        },
        "QueueName": "aws-ip-videoanalytics-VLControlPlane-forwardingRulesQueue",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "awsipvideoanalyticsVLControlPlaneforwardingRulesQueueStreamDLQD1B647B3",
              "Arn",
            ],
          },
          "maxReceiveCount": 5,
        },
        "VisibilityTimeout": 360,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ForwardingRulesQueuePolicy2F55AAA9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "SQS:SendMessage",
              "Condition": {
                "ArnLike": {
                  "aws:SourceArn": "arn:aws:s3:*:*:videoanalytics-fwd-rules*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "s3.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "ForwardingRulesQueueCCFF9CDD",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "ForwardingRulesQueueCCFF9CDD",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "ForwardingRulesTableE57C3B49": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "CustomerAccountId",
            "AttributeType": "S",
          },
          {
            "AttributeName": "ForwardingRulesName",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "DeletionProtectionEnabled": true,
        "KeySchema": [
          {
            "AttributeName": "CustomerAccountId",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "ForwardingRulesName",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableClass": "STANDARD",
        "TableName": "ForwardingRulesTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "awsipvideoanalyticsVLControlPlaneforwardingRulesQueueStreamDLQD1B647B3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "awsipvideoanalyticsVLControlPlaneforwardingRulesQueuekmsKeyA3AA2F57",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "awsipvideoanalyticsVLControlPlaneforwardingRulesQueuekmsKeyA3AA2F57": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789012:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:GenerateDataKey",
                "kms:Decrypt",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "s3.amazonaws.com",
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
